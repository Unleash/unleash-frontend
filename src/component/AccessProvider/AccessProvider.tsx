import { FC } from "react";

import AccessContext from '../../contexts/AccessContext'
import { ADMIN } from "./permissions";

// TODO: Type up redux store
interface IAccessProvider {
    store: any;
}

interface IPermission {
  permission: string;
  project: string | null;
}

const AccessProvider: FC<IAccessProvider> = ({store, children}) => {
  const hasAccess = (permission: string, project: string) => {
    const permissions = store.getState().user.get('permissions') || [];

    const result = permissions.some((p: IPermission) => {
      if(p.permission === ADMIN) {
        return true
      }
      if(p.permission === permission && p.project === project) {
        return true;
      }
      return false;
    });

    return result;
  };

  const context = { hasAccess };

  return <AccessContext.Provider value={context}>{children}</AccessContext.Provider>
}

export default AccessProvider;