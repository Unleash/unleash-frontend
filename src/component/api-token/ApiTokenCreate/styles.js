import { makeStyles } from '@material-ui/styles';

export const useStyles = makeStyles({
    addApiKeyForm: {
        display: 'flex',
        flexDirection: 'column',
    },
});
