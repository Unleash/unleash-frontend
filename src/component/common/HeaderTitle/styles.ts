import makeStyles from '@mui/styles/makeStyles';

export const useStyles = makeStyles(theme => ({
    headerTitleContainer: {
        display: 'flex',
        justifyContent: 'space-between',
        alignItems: 'center',
        position: 'relative',
    },
    headerTitle: {
        fontSize: theme.fontSizes.mainHeader,
        fontWeight: 'normal',
    },
    headerActions: {
        position: 'absolute',
        right: 0,
    },
}));
