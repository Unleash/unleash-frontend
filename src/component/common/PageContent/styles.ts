import { makeStyles } from 'tss-react/mui';

export const useStyles = makeStyles()(theme => ({
    headerContainer: {
        padding: theme.spacing(2, 4),
        borderBottom: `1px solid ${theme.palette.grey[100]}`,
        [theme.breakpoints.down('md')]: {
            padding: '1.5rem 1rem',
        },
    },
    bodyContainer: {
        padding: '2rem',
        [theme.breakpoints.down('md')]: {
            padding: '1rem',
        },
    },
    paddingDisabled: {
        padding: '0',
    },
    borderDisabled: {
        border: 'none',
    },
}));
