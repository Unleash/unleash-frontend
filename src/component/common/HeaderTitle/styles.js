import { makeStyles } from '@material-ui/styles';

export const useStyles = makeStyles(theme => ({
    headerTitleContainer: {
        display: 'flex',
        justifyContent: 'space-between',
        alignItems: 'center',
        position: 'relative',
    },
    headerTitle: {
        fontSize: theme.fontSizes.mainHeader,
        fontWeight: 500,
    },
    headerActions: {
        position: 'absolute',
        right: 0,
    },
}));
