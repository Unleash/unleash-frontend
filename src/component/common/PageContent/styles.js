import { makeStyles } from '@material-ui/styles';

export const useStyles = makeStyles(theme => ({
    headerContainer: {
        padding: theme.padding.pageContent.header,
        borderBottom: theme.borders.default,
    },
    bodyContainer: {
        padding: theme.padding.pageContent.body,
    },
}));
