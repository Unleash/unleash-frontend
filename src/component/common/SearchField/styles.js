import { makeStyles } from "@material-ui/styles";

export const useStyles = makeStyles(theme => ({
    search: {
        display: "flex",
        alignItems: "center",
        backgroundColor: "#fff",
        borderRadius: "3px",
        padding: "0.25rem 0.5rem",
        maxWidth: "450px"
    },
    searchIcon: {
        marginRight: "8px"
    },
    inputRoot: {
        width: "100%"
    }
}));
