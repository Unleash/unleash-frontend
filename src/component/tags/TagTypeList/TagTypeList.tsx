import { useContext, useState } from 'react';
import PropTypes from 'prop-types';
import { Link, useHistory } from 'react-router-dom';
import {
    Button,
    IconButton,
    List,
    ListItem,
    ListItemIcon,
    ListItemText,
    Tooltip,
} from '@material-ui/core';
import { Add, Delete, Edit, Label } from '@material-ui/icons';
import HeaderTitle from 'component/common/HeaderTitle';
import PageContent from 'component/common/PageContent/PageContent';
import ConditionallyRender from 'component/common/ConditionallyRender/ConditionallyRender';
import {
    DELETE_TAG_TYPE,
    UPDATE_TAG_TYPE,
} from 'component/providers/AccessProvider/permissions';
import Dialogue from 'component/common/Dialogue/Dialogue';
import useMediaQuery from '@material-ui/core/useMediaQuery';
import styles from './TagTypeList.module.scss';
import AccessContext from 'contexts/AccessContext';
import useTagTypesApi from 'hooks/api/actions/useTagTypesApi/useTagTypesApi';
import useTagTypes from 'hooks/api/getters/useTagTypes/useTagTypes';
import useToast from 'hooks/useToast';
import PermissionIconButton from 'component/common/PermissionIconButton/PermissionIconButton';
import { formatUnknownError } from 'utils/formatUnknownError';
import { ITagType } from 'interfaces/tags';

export const TagTypeList = () => {
    const { hasAccess } = useContext(AccessContext);
    const [deletion, setDeletion] = useState<{
        open: boolean;
        name?: string;
    }>({ open: false });
    const history = useHistory();
    const smallScreen = useMediaQuery('(max-width:700px)');
    const { deleteTagType } = useTagTypesApi();
    const { tagTypes, refetch } = useTagTypes();
    const { setToastData, setToastApiError } = useToast();

    const deleteTag = async () => {
        try {
            if (deletion.name) {
                await deleteTagType(deletion.name);
                refetch();
                setDeletion({ open: false });
                setToastData({
                    type: 'success',
                    show: true,
                    title: 'Successfully deleted tag type.',
                });
            }
        } catch (error) {
            setToastApiError(formatUnknownError(error));
        }
    };

    let header = (
        <HeaderTitle
            title="Tag Types"
            actions={
                <ConditionallyRender
                    condition={hasAccess(UPDATE_TAG_TYPE)}
                    show={
                        <ConditionallyRender
                            condition={smallScreen}
                            show={
                                <Tooltip title="Add tag type">
                                    <IconButton
                                        aria-label="add tag type"
                                        onClick={() =>
                                            history.push('/tag-types/create')
                                        }
                                    >
                                        <Add />
                                    </IconButton>
                                </Tooltip>
                            }
                            elseShow={
                                <Button
                                    variant="contained"
                                    color="primary"
                                    onClick={() =>
                                        history.push('/tag-types/create')
                                    }
                                >
                                    New tag type
                                </Button>
                            }
                        />
                    }
                />
            }
        />
    );

    const renderTagType = (tagType: ITagType) => {
        let link = (
            <Link to={`/tag-types/edit/${tagType.name}`}>
                <strong>{tagType.name}</strong>
            </Link>
        );
        let deleteButton = (
            <Tooltip title={`Delete ${tagType.name}`}>
                <IconButton
                    onClick={() =>
                        setDeletion({
                            open: true,
                            name: tagType.name,
                        })
                    }
                >
                    <Delete />
                </IconButton>
            </Tooltip>
        );

        return (
            <ListItem
                key={`${tagType.name}`}
                classes={{ root: styles.tagListItem }}
            >
                <ListItemIcon>
                    <Label />
                </ListItemIcon>
                <ListItemText primary={link} secondary={tagType.description} />
                <PermissionIconButton
                    permission={UPDATE_TAG_TYPE}
                    // @ts-expect-error
                    component={Link}
                    to={`/tag-types/edit/${tagType.name}`}
                >
                    <Edit className={styles.icon} titleAccess="Edit tag type" />
                </PermissionIconButton>
                <ConditionallyRender
                    condition={hasAccess(DELETE_TAG_TYPE)}
                    show={deleteButton}
                />
            </ListItem>
        );
    };
    return (
        <PageContent headerContent={header}>
            <List>
                <ConditionallyRender
                    condition={tagTypes.length > 0}
                    show={tagTypes.map(tagType => renderTagType(tagType))}
                    elseShow={<ListItem>No entries</ListItem>}
                />
            </List>
            <Dialogue
                title="Really delete Tag type?"
                open={deletion.open}
                onClick={deleteTag}
                onClose={() => {
                    setDeletion({ open: false });
                }}
            />
        </PageContent>
    );
};

TagTypeList.propTypes = {
    tagTypes: PropTypes.array.isRequired,
    fetchTagTypes: PropTypes.func.isRequired,
    removeTagType: PropTypes.func.isRequired,
};
