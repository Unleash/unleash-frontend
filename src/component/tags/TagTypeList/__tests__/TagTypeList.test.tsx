import React from 'react';
import { TagTypeList } from 'component/tags/TagTypeList/TagTypeList';
import renderer from 'react-test-renderer';
import { MemoryRouter } from 'react-router-dom';
import { ThemeProvider, Theme, StyledEngineProvider } from '@mui/styles';
import theme from 'themes/mainTheme';
import AccessProvider from 'component/providers/AccessProvider/AccessProvider';
import { ADMIN } from 'component/providers/AccessProvider/permissions';
import UIProvider from 'component/providers/UIProvider/UIProvider';


declare module '@mui/styles/defaultTheme' {
  // eslint-disable-next-line @typescript-eslint/no-empty-interface
  interface DefaultTheme extends Theme {}
}


test('renders an empty list correctly', () => {
    const tree = renderer.create(
        <MemoryRouter>
            <StyledEngineProvider injectFirst>
                <ThemeProvider theme={theme}>
                    <UIProvider>
                        <AccessProvider permissions={[{ permission: ADMIN }]}>
                            <TagTypeList />
                        </AccessProvider>
                    </UIProvider>
                </ThemeProvider>
            </StyledEngineProvider>
        </MemoryRouter>
    );
    expect(tree).toMatchSnapshot();
});
