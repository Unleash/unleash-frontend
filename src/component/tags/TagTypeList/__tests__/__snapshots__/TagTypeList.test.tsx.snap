// Vitest Snapshot v1

exports[`renders an empty list correctly 1`] = `
[
  <div>
    <div
      className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 tss-15wj2kz-container mui-177gdp-MuiPaper-root"
    >
      <div
        className="tss-1ywhhai-headerContainer"
      >
        <div
          className="tss-1ylehva-headerContainer"
        >
          <div
            className="tss-1uxyh7x-topContainer"
          >
            <div
              className="tss-sd6bs4-header"
              data-loading={true}
            >
              <h1
                className="MuiTypography-root MuiTypography-h1 tss-7lbvh4-headerTitle mui-ylrecv-MuiTypography-root"
              >
                Tag Types
              </h1>
            </div>
            <div
              className="tss-u5t8ea-headerActions"
            >
              <button
                className="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  mui-1aw3qf3-MuiButtonBase-root-MuiButton-root"
                disabled={false}
                onBlur={[Function]}
                onClick={[Function]}
                onContextMenu={[Function]}
                onDragLeave={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                tabIndex={0}
                type="button"
              >
                New tag type
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        className="tss-54jt3w-bodyContainer"
      >
        <ul
          className="MuiList-root MuiList-padding mui-h4y409-MuiList-root"
        >
          <li
            className="MuiListItem-root MuiListItem-gutters MuiListItem-padding mui-vlytkl-MuiListItem-root"
            disabled={false}
          >
            No entries
          </li>
        </ul>
      </div>
    </div>
  </div>,
  <div
    aria-atomic={true}
    aria-live="polite"
    className="tss-i8rqz1-container"
    data-testid="ANNOUNCER_ELEMENT_TEST_ID"
    role="status"
  />,
]
`;
